infrastructure/src/main/java/org/corfudb/infrastructure/log/StreamLogFiles.java

```diff
@@ -1275,7 +1281,7 @@ public void reset() {
         dataStore.resetStartingAddress();
         dataStore.resetTailSegment();
         logMetadata = new LogMetadata();
-        writeChannels.clear();
```

Comment:
Maybe your test can't trigger the issue because of this. Why was this dropped?

Reply:
Could you please explain your question?
It was dropped because if you clean this you will have FD leaks

Reply:
But if you have the locks protecting it, why adding this line could have the leaks? Does that mean the lock doesn't work at all?

Reply:
The leak was happening due to a synchronization problem between when `writeChannels.clear` and `writeChannels.computeIfAbsent` , but the RW lock is suppose to solve this synchronization issue. So, I don't understand why `writeChannels.clear();` has been removed. 

IF there are 4 segment handles [s1, s2, s3, s4] and the compaction mark is at 0, then a reset happens, then the same segment handles will be used to populate the state transfer data, which is incorrect. Essentially, your keeping the file channel handles in the map, but deleting the files underneath. What am I missing ?

Reply:
btw closeSegmentHandlers() will also remove the handle from the wirteChannels

Reply:
Ah, I see.

Reply:
But it still doesn't explain why removing this line could cause test to fail when lock is being added...

Reply:
I was asking @xnull , I didn't see the test failing without the `writeschannel.clear` call. I didn't try it myself.

Reply:
@Maithem I can't say for sure, my guess is somehow new segments get added after we execute `closeSegmentHandlers` and those segments get cleaned afterward by `writeSegments.clean()`.
I reproduced FD leaks one more time right now when I added epoch checks, and unfortunately, it has leaks.
Cleaning writeChannels prevents us from leaks.

P.S. one small question - https://github.com/CorfuDB/CorfuDB/blob/00f40e2e700313ca422a7f03f8fc92da58e0b67b/infrastructure/src/main/java/org/corfudb/infrastructure/LogUnitServer.java#L158

should it be check for `serverContext.getLogUnitEpochWaterMark()` ?

Reply:
Just to be consistent `getLogUnitEpochWaterMark` should be used, because that's what the BatchProcessor uses.

Reply:
ok, thanks